react_basic/src/pages/Login.jsx

```diff
@@ -0,0 +1,193 @@
+import { useState, useRef, useContext } from 'react';
+import { useNavigate, Link } from 'react-router-dom';
+import { PageWrapperContext } from '../pages/PageWrapper.jsx';
+import Loader from '../components/Loader.jsx';
+import validationService from '../services/validationService.js';
+import userService from '../services/userService.js';
+import '../styles/Login.css';
+
+function Login({ isLogin }) {
+  const navigate = useNavigate();
+
+  const {
+    setUser,
+    setErrorMsg,
+    setSuccessMsg,
+    setSuccessMsgCallback,
+    setSuccessMsgMillis,
+  } = useContext(PageWrapperContext);
+
+  const [values, setValues] = useState({ email: '', password: '' });
+  const [errors, setErrors] = useState({});
+
+  const [isLoading, setIsLoading] = useState(false);
+  const [isPasswordShown, setIsPasswordShown] = useState(false);
+
+  const loginFormRef = useRef();
+  const passwordInputRef = useRef();
+
+  function successCallback() {
+    navigate('/blog');
+  }
+
+  const title = isLogin ? 'Log in' : 'Sign up';
+
+  const msg = isLogin
+    ? "Don't have an account yet? Please, click "
+    : 'Have an account already? Please, click ';
+
+  function handleValueChange(e) {
+    const input = e.target;
+    setValues((prev) => ({ ...prev, [input.name]: input.value }));
+  }
+
+  async function handleSubmit(e) {
+    e.preventDefault();
+
+    setIsLoading(true);
+
+    try {
+      const user = await validationService.validateLoginForm(
+        loginFormRef.current,
+        isLogin,
+        setErrors,
+      );
+
+      if (loginFormRef.current.checkValidity()) {
+        if (isLogin) {
+          setUser(user);
+          setSuccessMsg("You've logged in successfully!");
+          setSuccessMsgCallback(successCallback);
+          setSuccessMsgMillis(3000);
+        } else {
+          const newUser = await userService.createUser(
+            values.email,
+            values.password,
+          );
+
+          setUser(newUser);
+          setSuccessMsg("You've created a user successfully!");
+          setSuccessMsgCallback(successCallback);
+          setSuccessMsgMillis(3000);
+        }
+      }
+    } catch {
+      setErrorMsg(`Error ${isLogin ? 'authenticating' : 'creating'} a user!`);
+    } finally {
+      setIsLoading(false);
+    }
+  }
+
+  function togglePasswordVisibility() {
```

Comment:
Not sure about this function, from my side of view it's easier to change input type directly in input: 
`<input type={isPasswordShown ? 'text' : 'password'}>`
it's more React approach